dll/win32/kernel32/kernel32_vista/vista.c

```diff
@@ -58,22 +99,78 @@ QueryFullProcessImageNameW(HANDLE hProcess,
                                            &Needed);
         Result = DynamicBuffer;
     }
-    else Result = (PUNICODE_STRING)Buffer;
-
-    if (!NT_SUCCESS(Status)) goto Cleanup;
+    else
+    {
+        Result = (PUNICODE_STRING)Buffer;
+    }
 
-    if (Result->Length / sizeof(WCHAR) + 1 > *pdwSize)
+    if (!NT_SUCCESS(Status))
     {
-        Status = STATUS_BUFFER_TOO_SMALL;
         goto Cleanup;
     }
 
-    *pdwSize = Result->Length / sizeof(WCHAR);
-    memcpy(lpExeName, Result->Buffer, Result->Length);
-    lpExeName[*pdwSize] = 0;
+    /* The name should use the Win32 path format. */
+    if (dwFlags == 0)
+    {
+        pszResult = RtlAllocateHeap(RtlGetProcessHeap(), 0, Result->Length + sizeof(WCHAR));
+
+        if (!pszResult)
+        {
+            Status = STATUS_NO_MEMORY;
+            goto Cleanup;
+        }
+
+        RtlStringCbCopyNW(pszResult, Result->Length + sizeof(WCHAR), Result->Buffer, Result->Length);
+
+        if (!DevicePathToDosPath(pszResult, Result->Length + sizeof(WCHAR)))
```

Comment:
In order to fix what Mark pointed out, you would need to combine both the copy and the `DevicePathToDosPath` call together, with a modified function prototype for the latter, so that it could be used as follows instead:
`DevicePathToDosPath(pszResult, Result->Length + sizeof(WCHAR), Result->Buffer, Result->Length)`